### DISCUSSION TITLE
Lost state when calling `router.refresh()` in app directory

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#44056] [Help]
Hello,

I am encountering the loss of client side state when applying the mutation pattern in the `app` directory described here:
https://beta.nextjs.org/docs/data-fetching/mutating

In the app directory under `app/[id]/page.tsx` I import a client component that has a form for editing the name of a resource.
The code of that client component looks like this.

```jsx
    const router = useRouter();

    const [success, setSuccess] = useState(false);

    const onSubmit = async (e: SyntheticEvent) => {
        e.preventDefault();
        const target = e.target as typeof e.target & {
            name: { value: string };
        };
        const data = {
            name: target.name.value,
        };
        await saveAction(data);
        setSuccess(true);
        router.refresh()
    }
```

When I entered that page via soft navigation (Next Link or router.push) the first time when I then call `router.refresh()` the success state will be lost and is `false` again when `router.refresh()` completed. Subsequent calls of `router.refresh()` do not lead to that lost success state which then remains `true` as originally intended. 

When I enter the page via a hard reload in the browser this problem is not present. Also when I entered the page via calling sth. like the below function this problem does not occur.

```jsx
   const navigate = () => {
        startTransition(() => {
            router.push("/dynamic-id");
            router.refresh();
        })
    }

```

So for me this appears to be related to using `router.refresh()` for the first time on a page that was entered via soft navigation.

Here is a link to the bug report I wrote:
https://github.com/vercel/next.js/issues/43650

Here is a link to the demo project I wrote for reproducing this problem:
https://github.com/schimi-dev/router-refresh-bug-report

Did anyone encounter this or a similar problem while expermenting with the app directory? Any explainations or suggestions?

Best regards,

Michael

**Comment by akomm** [comment]
same issue on my side.
